Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h

```diff
@@ -0,0 +1,78 @@
+#pragma once
+#ifndef Geometry_TrackerGeometryBuilder_phase1PixelTopology_h
+#define Geometry_TrackerGeometryBuilder_phase1PixelTopology_h
+
+#include <cstdint>
+
+namespace phase1PixelTopology {
+
+  constexpr uint16_t numRowsInRoc     = 80;
```

Comment:
@fwyzard - shouldn't these hardcoded numbers be taken from an EventSetup nowadays?

Reply:
NO!
they are hardcoded for good reason: notice this is for phase1 and we have ONE and only ONE
phase1 pixel detector.

This is supposed to be used in very high performance computing. NO runtime ifs.
Only hardcoded constants.

Reply:
@VinInn - the constants are iov dependant, the same as other conditions.

Reply:
constants are constants. please try to be realistic.
THESE numbers do not change, will never change. Are hardwired in the hardware down in the pit.
and in any case are Hardwired also in current code see for instance
https://cmssdt.cern.ch/dxr/CMSSW/source/Geometry/TrackerGeometryBuilder/src/RectangularPixelTopology.cc#85


The correct way of handling all this even with iov is to have all these numbers hardwired in their namespace as here, have High Performance algorithms templeted with that and choose using IOV and having the Low Performance analyzer (DQM) use a generic wrapper with ifs all over the place (as done in TrackerTopology)
TrackerTopology is not the way to go to get high performance

Reply:
to be clear, this file is just a HighPerformanceVersion of RectangularPixelTopology not much more hardwired only clearer actually while the "geometry constants" in RectangularPixelTopology are spread all over the place...

Reply:
How will these constant relate to PTrackerParameters?

Reply:
Do you expect to modify PTrackerParameters?
In any case they relate to the real detector, better PTrackerParameters do the same.
So they relate to  PTrackerParameters as they originate from the very same source.
And again this file is supposed to emulate RectangularPixelTopology producing the very same results for Phase1 detector

Reply:
IMHO, there should not be two ways to retrieve the same information. If this new high performant approach will be used based on IOV, perhaps, it should replace the PTrackerParameters.

Reply:
Sure, at some point. When we will have enough person power.
For the time being we are in R&D and we do not have the resources to fix legacy code.
CMS will need to decide, at some point, what to do will legacy code once we move to heterogeneous high-performance-coding for the upgrades.
Most probably we will need to maintain two (more?) sources of truth (and, as already pointed out with RectangularPixelTopology example, the current code is indeed full of multiple sources of truth)